File: frontend/src/types/enums/gap-severity.ts
export type GapSeverity = 'minor' | 'major' | 'critical'
export type GapState = 'detected' | 'reviewed' | 'accepted' | 'false_positive' | 'resurveyed' | 'closed'
export const GAP_SEVERITY_LABEL: Record<GapSeverity,string> = { minor: '轻微', major: '主要', critical: '严重' }
export const GAP_STATE_LABEL: Record<GapState,string> = { detected:'待复核', reviewed:'已复核', accepted:'接受补测', false_positive:'误报', resurveyed:'已补测', closed:'已关闭' }
/** 可参与多快照补测调度的缺口状态：待复核或已复核。 */
export const RESURVEY_SELECTABLE_STATES: GapState[] = ['detected', 'reviewed']
export const isResurveySelectable = (state: GapState): boolean => RESURVEY_SELECTABLE_STATES.includes(state)
export const GAP_SEVERITY_ORDER: Record<GapSeverity,number> = { critical: 3, major: 2, minor: 1 }

File: backend/internal/constants/gap_severity.go
package constants

import "fmt"

type GapSeverity string
type GapState string

const (
	SeverityMinor    GapSeverity = "minor"
	SeverityMajor    GapSeverity = "major"
	SeverityCritical GapSeverity = "critical"

	GapDetected      GapState = "detected"
	GapReviewed      GapState = "reviewed"
	GapAccepted      GapState = "accepted"
	GapFalsePositive GapState = "false_positive"
	GapResurveyed    GapState = "resurveyed"
	GapClosed        GapState = "closed"
)

var gapTransitions = map[GapState]map[GapState]struct{}{
	GapDetected:      {GapReviewed: {}},
	GapReviewed:      {GapAccepted: {}, GapFalsePositive: {}},
	GapAccepted:      {GapResurveyed: {}},
	GapFalsePositive: {GapClosed: {}},
	GapResurveyed:    {GapClosed: {}},
	GapClosed:        {},
}

func (s GapState) Valid() bool {
	_, ok := gapTransitions[s]
	return ok
}

func (s GapState) CanTransition(target GapState) bool {
	allowed, ok := gapTransitions[s]
	if !ok {
		return false
	}
	_, ok = allowed[target]
	return ok
}

// Schedulable reports whether a gap snapshot may join a resurvey task sheet.
// Only snapshots awaiting review (detected) or already reviewed qualify.
func (s GapState) Schedulable() bool {
	return s == GapDetected || s == GapReviewed
}

// SeverityRank orders severities for resurvey scheduling; lower ranks schedule first.
func SeverityRank(value GapSeverity) int {
	switch value {
	case SeverityCritical:
		return 0
	case SeverityMajor:
		return 1
	default:
		return 2
	}
}

func ParseGapState(value string) (GapState, error) {
	state := GapState(value)
	if !state.Valid() {
		return "", fmt.Errorf("unknown gap state %q", value)
	}
	return state, nil
}

func SeverityForRatio(ratio float64) GapSeverity {
	switch {
	case ratio >= 0.12:
		return SeverityCritical
	case ratio >= 0.05:
		return SeverityMajor
	default:
		return SeverityMinor
	}
}
